iothub/device/src/Common/Security/CertificateInstaller.cs

```
@@ -0,0 +1,70 @@
+        /// See <see href="TODO (Sample ref) - Sindhu"/>
+        /// for more information.
+        /// </remarks>
+        /// <param name="certificates">The certificate chain to ensure is installed.</param>
+        public static void EnsureChainIsInstalled(X509Certificate2Collection certificates)
+        {
+            if (certificates == null || certificates.Count == 0)
+            {
+                if (Logging.IsEnabled) Logging.Info(null, $"{nameof(CertificateInstaller)} parameter 'certificates' was null or empty.");
+
+                return;
+            }
+
+            // Certificate install on Windows is a multi-step process, and in the case that someone might have more than one
+            // Device client we'll want to ensure these actions (get certs, install certs) are atomic.
+            lock (s_certOperationsLock)
+            {
+                var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser);
+
+                try
+                {
+                    store.Open(OpenFlags.ReadWrite);
+
+                    foreach (X509Certificate2 certificate in certificates)
+                    {
+                        X509Certificate2Collection results = store.Certificates.Find(
+                            X509FindType.FindByThumbprint,
+                            certificate.Thumbprint,
+                            false);
+                        if (results.Count == 0)
+                        {
+                            if (Logging.IsEnabled) Logging.Info(null, $"{nameof(CertificateInstaller)} adding cert with thumbprint {certificate.Thumbprint} to X509 store.");
+
+                            store.Add(certificate);
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    if (Logging.IsEnabled) Logging.Error(null, $"{nameof(CertificateInstaller)} failed to read or write to cert store due to: {ex}");
```

we are catching the exception and swallowing it, isn't that considered a broken scenario? do we just continue?

---

I am sure the auth will fail going forward. But good point. I was trying to keep the same code between DPS and DeviceClient but this might be worth looking into. I am not sure why we swallow it in the DPS scenario.

---

however, getting unauthorize without understanding why the provided certs didn't work is confusing, I'd rather expose the exception here or throw and catch it in the caller and throw a meaningful exception.

---

I think catching it in the caller is a good idea. I will do that.

---

But because this is a helper, it should not really throw exceptions. I'll see if I can do something in the caller. The best would be a generic catch only.